workers/unity/Packages/io.improbable.gdk.core/Collections/EntityRangeCollection.cs

```diff
@@ -0,0 +1,184 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Improbable.Gdk.Core
+{
+    public class EntityRangeCollection : IEnumerable<EntityId>
+    {
+        private EntityIdRange firstElement;
+        private readonly Queue<EntityIdRange> queue = new Queue<EntityIdRange>();
+        private int version = 0;
+
+        public uint Count { get; private set; } = 0;
+
+        public EntityRangeCollection Take(uint count)
+        {
+            if (count > Count)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count), count, "Collection does not have enough ids stored");
+            }
+
+            var collection = new EntityRangeCollection();
+            while (count > 0)
+            {
+                if (firstElement.Count <= count)
+                {
+                    count -= firstElement.Count;
+                    collection.Add(firstElement);
+                    firstElement = queue.Count > 0
+                        ? queue.Dequeue()
+                        : new EntityIdRange();
+                }
+                else
+                {
+                    var splitRange = new EntityIdRange(firstElement.FirstEntityId, count);
+                    collection.Add(splitRange);
+                    firstElement.FirstEntityId = new EntityId(firstElement.FirstEntityId.Id + count);
+                    firstElement.Count -= count;
+                    break;
+                }
+            }
+
+            Count -= count;
+            version++;
+
+            return collection;
+        }
+
+        internal void Add(EntityIdRange range)
+        {
+            if (!range.FirstEntityId.IsValid())
+            {
+                throw new ArgumentException("Invalid EntityId in range", nameof(range));
+            }
+
+            if (range.Count == 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(range), "EntityId range cannot be 0 in length.");
+            }
+
+            if (firstElement.Count == 0)
+            {
+                firstElement = range;
+            }
+            else
+            {
+                queue.Enqueue(range);
+            }
+
+            Count += range.Count;
+            version++;
+        }
+
+        public void Clear()
+        {
+            firstElement = new EntityIdRange();
+            queue.Clear();
+            Count = 0;
+            version++;
+        }
+
+        internal struct EntityIdRange
+        {
+            public EntityId FirstEntityId;
```

Comment:
Maybe deal with `long`s until the public API layer for simplicity?

Reply:
I'd rather leave it like this in case we get more restrictions on EntityId's.
Currently this lets me use `IsValid` to check for <= 0 id's, layer we might check for the <= 62^2-1 range as well.